--- a/dlls/oleaut32/vartype.c
+++ b/dlls/oleaut32/vartype.c
@@ -3496,7 +3496,7 @@
  */
 HRESULT WINAPI VarCyFromR8(double dblIn, CY* pCyOut)
 {
-#if defined(__GNUC__) && (defined(__i386__) || defined(__x86_64__))
+#if defined(__i386__) || (defined(__x86_64__) && !defined(__arm64ec__))
   /* This code gives identical results to Win32 on Intel.
    * Here we use fp exceptions to catch overflows when storing the value.
    */
--- a/dlls/vcomp/main.c
+++ b/dlls/vcomp/main.c
@@ -135,7 +135,7 @@
         args[i] = va_arg(valist, void *);
 }
 
-#if defined(__GNUC__) && (defined(__i386__) || defined(__x86_64__))
+#if (defined(__i386__) || defined(__x86_64__)) && !defined(__arm64ec__)
 
 static inline char interlocked_cmpxchg8(char *dest, char xchg, char compare)
 {
--- a/dlls/win32u/dibdrv/primitives.c
+++ b/dlls/win32u/dibdrv/primitives.c
@@ -246,7 +246,7 @@
 
 static inline void memset_32( DWORD *start, DWORD val, DWORD size )
 {
-#if defined(__GNUC__) && (defined(__i386__) || defined(__x86_64__))
+#if defined(__GNUC__) && (defined(__i386__) || defined(__x86_64__)) && !defined(__arm64ec__)
     DWORD dummy;
     __asm__ __volatile__( "cld; rep; stosl"
                           : "=c" (dummy), "=D" (dummy)
@@ -258,7 +258,7 @@
 
 static inline void memset_16( WORD *start, WORD val, DWORD size )
 {
-#if defined(__GNUC__) && (defined(__i386__) || defined(__x86_64__))
+#if defined(__GNUC__) && (defined(__i386__) || defined(__x86_64__)) && !defined(__arm64ec__)
     DWORD dummy;
     __asm__ __volatile__( "cld; rep; stosw"
                           : "=c" (dummy), "=D" (dummy)
